pallets/file-system/src/lib.rs

```
@@ -0,0 +1,290 @@
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn storage_requests)]
+    pub type StorageRequests<T: Config> =
+        StorageMap<_, Blake2_128Concat, FileLocation<T>, StorageRequestMetadata<T>>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn storage_requests_expirations)]
+    pub type StorageRequestExpirations<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        BlockNumberFor<T>,
+        BoundedVec<FileLocation<T>, T::MaxExpiredStorageRequests>,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Notifies that a new file has been requested to be stored.
+        NewStorageRequest {
+            who: T::AccountId,
+            location: FileLocation<T>,
+            fingerprint: Fingerprint<T>,
+            size: StorageUnit<T>,
+            user_multiaddr: MultiAddress<T>,
+        },
+
+        /// Notifies that a BSP has been accepted to store a given file.
+        AcceptedBspVolunteer {
+            who: T::AccountId,
+            location: FileLocation<T>,
+            fingerprint: Fingerprint<T>,
+            bsp_multiaddress: MultiAddress<T>,
+        },
+
+        /// Notifies the expiration of a storage request.
+        StorageRequestExpired { location: FileLocation<T> },
+    }
+
+    // Errors inform users that something went wrong.
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Storage request already registered for the given file.
+        StorageRequestAlreadyRegistered,
+        /// Storage request not registered for the given file.
+        StorageRequestNotRegistered,
+        /// BSP already volunteered to store the given file.
+        BspVolunteerFailed,
+        /// BSP already confirmed to store the given file.
+        BspAlreadyConfirmed,
+        /// No slot available found in blocks to insert storage request expiration time.
+        StorageRequestExpiredNoSlotAvailable,
```

Actually, this error should never happen

---

In the off chance the storage request expiry fails to append to a `BoundedVec`, we need to return an error from the extrinsic, and it still should tell you what happened (even if it will never happen xD)

---

Sounds reasonable. In that case I'd use the `expect_or_err` macro so that in tests this panics with unreachable, but in wasm it reverts with that error.

---

good idea